mne/tests/test_evoked.py

```diff
@@ -252,6 +261,23 @@ def test_shift_time_evoked():
     assert_allclose(ave_normal.data, ave_absolute.data, atol=1e-16, rtol=1e-3)
     assert_equal(ave_absolute.first, int(-0.3 * ave.info['sfreq']))
 
+    # subsample shift
+    shift = 1e-6  # 1 μs, should be well below 1/sfreq
+    ave = read_evokeds(fname, 0)
+    times = ave.times
+    ave.shift_time(shift)
+    assert_allclose(times + shift, ave.times, atol=1e-16, rtol=1e-12)
+
+    # test handling of Evoked.first, Evoked.last
+    ave = read_evokeds(fname, 0)
+    first_last = np.array([ave.first, ave.last])
+    # should shift by 0 samples
+    ave.shift_time(1e-6)
+    assert_array_equal(first_last, np.array([ave.first, ave.last]))
+    # should shift by 57 samples
+    ave.shift_time(57. / ave.info['sfreq'])
+    assert_array_equal(first_last + 57, np.array([ave.first, ave.last]))
```

Comment:
I suspect that something bad is going to happen on I/O round-trip following a shift, specifically for `Evoked` files. We always write `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`, but never `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES`:
```
            # First and last sample
            write_int(fid, FIFF.FIFF_FIRST_SAMPLE, e.first)
            write_int(fid, FIFF.FIFF_LAST_SAMPLE, e.last)
```
Then on read our decision about times goes:
```
        if first is not None:
            nsamp = last - first + 1
        elif first_time is not None:
            first = int(round(first_time * info['sfreq']))
            last = first + nsamp
    ...
    times = np.arange(first, last + 1, dtype=np.float) / info['sfreq']
```

In the spirit of TDD, can you confirm this by adding a test I/O round-tripping your `1e-6`-shifted `evoked` instance will remove the shift?

To fix this, we can instead:

1. Always also write `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES` (in addition to / keeping `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`
2. Prefer to use `FIFF_FIRST_TIME` and `FIFF_NO_SAMPLES`, if available, and if not available, fall back to `FIFF_FIRST_SAMPLE` and `FIFF_LAST_SAMPLE`

This should be backward compatible but also work for your sub-sample shift case.

Reply:
@larsoner see if you're satisfied with the new read logic and the updated tests.  Times are stable through save/load roundtrip up to 8 decimal places (in my ad hoc testing sometimes 9, never 10).  Seems good enough to me but I trust your judgment here.

Reply:
Yes looks good. But I realized there is a tiny simplification / DRY we can get from no longer passing `first/last` out of `_read_evoked`, pushing a commit now. See if you think it's good then we should be ready for @agramfort to look and merge.

I also pushed a commit to fix a failing test by specifying `atol`.